gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -674,62 +679,43 @@ case class TransformPostOverrides(isAdaptiveContext: Boolean) extends Rule[Spark
   }
 }
 
-object InMemoryTableScanHelper {
-  private def isGlutenTableCacheInternal(i: InMemoryTableScanExec): Boolean = {
-    // `ColumnarCachedBatchSerializer` is at velox module, so use class name here
-    i.relation.cacheBuilder.serializer.getClass.getSimpleName == "ColumnarCachedBatchSerializer" &&
-    i.supportsColumnar
-  }
-
-  def isGlutenTableCache(plan: SparkPlan): Boolean = {
-    plan match {
-      case i: InMemoryTableScanExec =>
-        isGlutenTableCacheInternal(i)
-      case q: QueryStageExec =>
-        // Compatible with Spark3.5 `TableCacheQueryStage`
-        isGlutenTableCache(q.plan)
-      case _ => false
-    }
-  }
-}
-
 // This rule will try to add RowToColumnarExecBase and ColumnarToRowExec
-// to support vanilla columnar scan.
+// to support vanilla columnar operators.
 case class VanillaColumnarPlanOverrides(session: SparkSession) extends Rule[SparkPlan] {
   @transient private val planChangeLogger = new PlanChangeLogger[SparkPlan]()
 
   private def replaceWithVanillaColumnarToRow(plan: SparkPlan): SparkPlan = plan match {
-    case c2r: ColumnarToRowExecBase if isVanillaColumnarReader(c2r.child) =>
+    case c2r: ColumnarToRowExecBase if PlanUtil.isVanillaColumnarOp(c2r.child) =>
       ColumnarToRowExec(c2r.child)
-    case c2r: ColumnarToRowExec if isVanillaColumnarReader(c2r.child) =>
+    case c2r: ColumnarToRowExec if PlanUtil.isVanillaColumnarOp(c2r.child) =>
       c2r
-    case _ if isVanillaColumnarReader(plan) =>
+    case _ if PlanUtil.isVanillaColumnarOp(plan) =>
       BackendsApiManager.getSparkPlanExecApiInstance.genRowToColumnarExec(ColumnarToRowExec(plan))
     case _ =>
       plan.withNewChildren(plan.children.map(replaceWithVanillaColumnarToRow))
   }
 
-  private def isVanillaColumnarReader(plan: SparkPlan): Boolean = plan match {
-    case _: BatchScanExec | _: DataSourceScanExec =>
-      !plan.isInstanceOf[GlutenPlan] && plan.supportsColumnar
-    case i: InMemoryTableScanExec =>
-      if (InMemoryTableScanHelper.isGlutenTableCache(i)) {
-        // `InMemoryTableScanExec` do not need extra RowToColumnar or ColumnarToRow
-        false
-      } else {
-        !plan.isInstanceOf[GlutenPlan] && plan.supportsColumnar
-      }
-    case _ => false
+  private def replaceWithVanillaRowToColumnar(plan: SparkPlan): SparkPlan = plan match {
+    case _ if PlanUtil.isVanillaColumnarOp(plan) =>
+      plan.withNewChildren(plan.children.map {
+        c =>
+          val child = replaceWithVanillaRowToColumnar(c)
+          if (child.isInstanceOf[GlutenPlan]) {
```

Comment:
If there is `PlanUtil.isVanillaColumnarOp`, should we add `PlanUtil.isGlutenColumnarOp`?